rfcs/0000-state-separation.md

```diff
@@ -0,0 +1,706 @@
+
+Here is a list of some of the signatures of `StateHandler`:
+
+```rust
+trait StateHandler<S, E> {
+    fn on_start(&mut self, _: &mut World) {
+    }
+
+    fn update(&mut self, _: &mut World) -> Trans<S> {
+        Trans::None
+    }
+
+    fn handle_event(&mut self, _: &mut World, _: &E) -> Trans<S> {
+        Trans::None
+    }
+
+    // snip
+}
+```
+
+Most notably `Trans` has been simplified to only take the `State`, and we no longer receive a parameterized `StateData<T>`.
+Events are passed by reference, since they are passed to multiple handlers (shadow, global, state).
+
+This also means that we need to _associate_ a state with its corresponding handler.
+This is done when building the application like this:
+
+```rust
+let mut game = Application::build("./assets")?
+    .with_state(State::Main, MainState)?
+    .build(GameDataBuilder::default())?;
+
+game.run();
+```
+
+Note that there are a couple of types which come with `State` implementations:
+* `&'static str`
+* `u8`, `u32`, `u64`
+* `i8`, `i32`, `i64`
+* `()` (the empty tuple)
+
+Note that the last one can be used as a "default state" when you only have one in your game:
+
+```rust
+let mut game = Application::build("./assets")?
+    .with_state((), MainState)?
+    .build(GameDataBuilder::default())?;
+
+game.run();
+```
+
+The initial state is determined by its `Default` implementation. So `()` would be `()`, `&str` would be `""` (the empty string).
+
+It's also trivial to implement a HashMap-based state implementation as long as your state implements `Hash + PartialEq + Eq`:
+
+```rust
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct SomeStruct {
+    // snip
+}
+
+impl<E> State<E> for SomeStruct {
+    type Storage = MapStateStorage<Self, E>;
+}
+```
+
+For enum-based states there's an even more efficient alternative, which is covered in the next section.
+
+## The `State` derive
+
+The most natural way to model states is to use an enum with unit variants. Unit variants are variants which have no structure or tuple body, like this:
+
+```rust
+enum State {
+  Loading,
+  Main,
+}
+```
+
+To provide a `State` implementation for this enum you can use the `State` derive:
+
+```rust
+#[derive(State, Debug, Clone)]
+enum State {
+  Loading,
+  Main,
+}
+```
+
+This will provide the `State` implementation, and implement `Default` to return the first element.
+
+When using this derive, the Storage implementation is also specialized to an implementation like this:
+
+```rust
+struct State_Storage<E> {
+  f1: Option<Box<dyn amethyst::StateHandler<State, E>>>,
+  f2: Option<Box<dyn amethyst::StateHandler<State, E>>>,
+}
+
+impl<E> amethyst::StateStorage<State, E> for State_Storage<E> {
+    fn get_mut(&mut self, value: &State) -> Option<&mut Box<dyn amethyst::StateHandler<State, E>>> {
+        match *self {
+            State::Loading => self.f1.as_mut(),
+            State::Main => self.f2.as_mut(),
```

Comment:
API restrictions here are a tad loose, one could in theory push a state of `(State::Loading, Main)`.  We'd want to either prevent that kind of usage at compile time, or `panic!` when it happens.  Each `State` enum variant should be tied to exactly one `StateHandler` type in my opinion.

Alternatively, is having the ability to pair a `StateHandler` with multiple state enum variants an explicit goal of this RFC?  If so why?  What kind of designs would that enable that we aren't currently accommodating?

Reply:
This RFC doesn't really specify which handler is tied to what state. All of that is explicitly mapped at application creation time (`with_state`).

This does open up the door to generic handlers. But if a handler is not generic over the state it can only be tied to one state. This seems pretty natural.